File: antlr-projeto/ExprEvalListener.java
import java.util.Stack;

public class ExprEvalListener extends ExprBaseListener {

    private Stack<Double> stack = new Stack<>();

    public double getResult() {
        return stack.peek();
    }

    @Override
    public void exitNumber(ExprParser.NumberContext ctx) {
        double value = Double.parseDouble(ctx.getText());
        stack.push(value);
    }

    @Override
    public void exitParent(ExprParser.ParentContext ctx) {
        // O valor da expressão entre parênteses já está na pilha
    }

    @Override
    public void exitSomaSub(ExprParser.SomaSubContext ctx) {
        double right = stack.pop();
        double left = stack.pop();
        if (ctx.getChild(1).getText().equals("+")) {
            stack.push(left + right);
        } else {
            stack.push(left - right);
        }
    }

    @Override
    public void exitMultDiv(ExprParser.MultDivContext ctx) {
        double right = stack.pop();
        double left = stack.pop();
        if (ctx.getChild(1).getText().equals("*")) {
            stack.push(left * right);
        } else {
            stack.push(left / right);
        }
    }

    @Override
    public void exitPot(ExprParser.PotContext ctx) {
        double right = stack.pop();
        double left = stack.pop();
        stack.push(Math.pow(left, right));
    }

    @Override
    public void exitAbsExpr(ExprParser.AbsExprContext ctx) {
        double val = stack.pop();
        stack.push(Math.abs(val));
    }

    @Override
    public void exitFactExpr(ExprParser.FactExprContext ctx) {
        double val = stack.pop();
        int intVal = (int) val;
        double result = 1;
        for (int i = 2; i <= intVal; i++) {
            result *= i;
        }
        stack.push(result);
    }
}

File: antlr-projeto/python/ExprEvalListener.py
from ExprListener import ExprListener
import math

class ExprEvalListener(ExprListener):
    def __init__(self):
        self.stack = []

    def exitNumber(self, ctx):
        value = float(ctx.getText())
        self.stack.append(value)

    def exitParent(self, ctx):
        # Valor já processado
        pass

    def exitSomaSub(self, ctx):
        right = self.stack.pop()
        left = self.stack.pop()
        op = ctx.getChild(1).getText()
        self.stack.append(left + right if op == '+' else left - right)

    def exitMultDiv(self, ctx):
        right = self.stack.pop()
        left = self.stack.pop()
        op = ctx.getChild(1).getText()
        self.stack.append(left * right if op == '*' else left / right)

    def exitPot(self, ctx):
        right = self.stack.pop()
        left = self.stack.pop()
        self.stack.append(math.pow(left, right))

    def exitAbsExpr(self, ctx):
        val = self.stack.pop()
        self.stack.append(abs(val))

    def exitFactExpr(self, ctx):
        val = self.stack.pop()
        self.stack.append(math.factorial(int(val)))
